### PR TITLE
test(convert): add coverage for edge cases in convert/mod.rs

### PR DESCRIPTION
**Pull request by mitsuru**
## Summary

`crates/fulgur/src/convert/mod.rs` のライン カバレッジを **94.70% → 96.28%**（+1.58 pp, 13 ライン）改善するテストを追加します。

対象モジュールの選定理由: カバレッジ測定で最も未カバーライン数が多いファイル上位3つ（`multicol_layout.rs` 93.80%、`convert/mod.rs` 94.70%、`convert/inline_root.rs` 94.76%）の中で、既存の `build_drawables` テスト基盤を流用でき、純粋な防衛ガードとエッジケースのテストが最も書きやすいのが `convert/mod.rs` と判断しました。

## Changes

- `crates/fulgur/src/convert/mod.rs` に新規テストモジュール `edge_case_tests` を追加

新規テスト一覧:

| テスト名 | 対象 | カバーするライン |
|----------|------|-----------------|
| `deeply_nested_html_does_not_panic_and_truncates_at_max_depth` | `convert_node` / `walk_semantics` の `MAX_DOM_DEPTH` (512) ガード | 333, 401 |
| `empty_id_attribute_is_not_stored_in_block_entry` | `extract_block_id` が `id=""` を `None` として返すこと | 978 |
| `whitespace_only_id_attribute_is_not_stored` | `extract_block_id` が `id="   "` (空白のみ) を `None` として返すこと | 978（paragraph path） |
| `fulgur_debug_env_var_triggers_debug_tree_without_panic` | `FULGUR_DEBUG` 環境変数による `debug_print_tree` 実行パス | 167, 277–310 |

**意図的に除外したブランチ（理由付き）:**

- `get_text_color` / `get_text_decoration` のデフォルト返値（行 1109–1110, 1149–1150）: `brush.id` が存在しない DOM ノードを指す場合にのみ発動する防衛パス。正常な文書処理では到達不可能であり、この条件を再現するには `shape_paragraph_glyph_runs` 内部に直接介入する必要があるため除外。
- `record_transform` / `record_multicol_rule` の early return（行 577, 580, 604, 631, 647）: `get_node` / `primary_styles` が `None` を返す場合のガード。DOM の整合性が保たれている通常処理では到達しない。
- `shape_paragraph_glyph_runs` 内の `panic!` メッセージ（行 921–924）: Parley クラスタイテレータが枯渇した場合の不変条件違反パニック。通常使用では到達不可能。

## Test plan

- [x] `cargo test -p fulgur --lib` — 1908 passed; 0 failed
- [x] `cargo clippy -- -D warnings` — clean
- [x] `cargo fmt --check` — clean
- [x] カバレッジ再測定: `convert/mod.rs` 94.70% → 96.28%

---
_Generated by [Claude Code](https://claude.ai/code/session_01RnQom5HGVAeMj62mBsptDr)_

## Summary by CodeRabbit

* **テスト**
  * 深い DOM 構造を安全に処理し、深度制限を超えた要素が登録されないことを検証します。
  * 空または空白のみの `id` 属性が適切に除外され、未トリムの値も保存されないことを確認します。

### CODE REVIEW

File: crates/fulgur/src/convert/mod.rs [hunk 1]
```diff
@@ -1920,3 +1920,183 @@ mod utility_fn_tests {
...
+        let bad_block_ids: Vec<_> = d
+            .block_styles
+            .values()
+            .filter(|e| e.id.as_deref().map(|s| s.is_empty()) == Some(true))
+            .collect();
+        assert!(
+            bad_block_ids.is_empty(),
+            "whitespace-only id must not be stored in block entries; found {}",
+            bad_block_ids.len()
+        );
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/convert/mod.rs:2137]: 対応済みです (45f5a9a)。`is_empty()` を `trim().is_empty()` に修正しました。`whitespace_only_id_attribute_is_not_stored` のフィルタ 2 か所と `empty_id_attribute_is_not_stored_in_block_entry` のフィルタ 2 か所、計 4 か所すべてで適用しています。空白文字のみの id 文字列が保存されるような回帰が発生した場合も、テストが確実に検出できます。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/fulgur/src/convert/mod.rs [hunk 2]
```diff
@@ -1920,3 +1920,183 @@ mod utility_fn_tests {
...
+        // so a race only risks spurious stderr noise, not test failures.
+        let html = "<!DOCTYPE html><html><body><h1>debug</h1><p>text</p></body></html>";
+        unsafe {
+            std::env::set_var("FULGUR_DEBUG", "1");
+        }
+        let result = std::panic::catch_unwind(|| build_drawables(html));
+        unsafe {
+            std::env::remove_var("FULGUR_DEBUG");
+        }
+        assert!(result.is_ok(), "debug_print_tree must not panic");
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/convert/mod.rs]: 対応済みです (45f5a9a)。`fulgur_debug_env_var_triggers_debug_tree_without_panic` テストを削除しました。

Rust edition 2024 のマルチスレッドテストプロセス内での `std::env::set_var` 呼び出しは `unsafe` 要件を満たさず、同一プロセス内の他スレッドが任意の環境変数を同時に読み取る可能性があるため、安全に分離する方法がありません。`FULGUR_DEBUG` ゲートの行（167 行目・277〜310 行目）は subprocess を用いた分離なしにはテストできないと判断し、今回のカバレッジ改善対象から除外しました。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/fulgur/src/convert/mod.rs [hunk 3]
```diff
@@ -1920,3 +1920,183 @@ mod utility_fn_tests {
...
+        // debug_print_tree (lines 167, 277-310), which writes to stderr.
+        // This test verifies the path executes without panicking.
+        // Safety: set_var is unsafe in edition 2024 because it races with
+        // concurrent reads; this test must not run in parallel with other
+        // tests that read FULGUR_DEBUG. Cargo runs tests in the same
+        // process by default, but debug_print_tree only writes to stderr
+        // so a race only risks spurious stderr noise, not test failures.
+        let html = "<!DOCTYPE html><html><body><h1>debug</h1><p>text</p></body></html>";
+        unsafe {
+            std::env::set_var("FULGUR_DEBUG", "1");
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/convert/mod.rs]: 対応済みです (45f5a9a)。`fulgur_debug_env_var_triggers_debug_tree_without_panic` テストを削除しました。Rust edition 2024 のマルチスレッドテストプロセス内では `set_var` の安全性要件を満たせないため、サブプロセス分離なしにこのテストを保持することはできないと判断しています。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/fulgur/src/convert/mod.rs [hunk 4]
```diff
@@ -1920,3 +1920,183 @@ mod utility_fn_tests {
...
+        // id="   " (spaces only) also trims to empty and must return None.
+        // A <p> is an inline root, so its id lands in d.paragraphs.
+        let html = "<!DOCTYPE html><html><body><p id=\"   \">paragraph</p></body></html>";
+
+        let d = build_drawables(html);
+
+        let bad_para_ids: Vec<_> = d
+            .paragraphs
+            .values()
+            .filter(|e| e.id.as_deref().map(|s| s.is_empty()) == Some(true))
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/convert/mod.rs]: 対応済みです (45f5a9a)。`is_empty()` を `trim().is_empty()` に修正しました。`whitespace_only_id_attribute_is_not_stored` と `empty_id_attribute_is_not_stored_in_block_entry` の両テストで、空白文字のみの id 文字列を確実に無効として検出できます。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/fulgur/src/convert/mod.rs [hunk 5]
```diff
@@ -1920,3 +1920,183 @@ mod utility_fn_tests {
...
+            .stack_size(64 * 1024 * 1024) // 64 MiB: blitz/taffy recurse before fulgur's guard fires
+            .spawn(move || build_drawables(&html))
+            .expect("thread spawn")
+            .join()
+            .expect("thread did not panic");
+
+        // At least the shallow entries (before the depth cap) must be recorded.
+        assert!(
+            !d.block_styles.is_empty(),
+            "shallow block entries must be recorded even with deep nesting"
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/convert/mod.rs:2016]: 対応済みです (45f5a9a)。深さインデックス 550 番目の `<div>` に `id="beyond-depth-limit"` を付与し、その要素が `block_styles` と `paragraphs` の**両方**に存在しないことを `assert!` で検証するよう強化しました。深さガードを除去した場合はこのアサーションが失敗するため、ガードの実際の発火を確認できます。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/fulgur/src/convert/mod.rs [hunk 6]
```diff
@@ -1920,3 +1920,185 @@ mod utility_fn_tests {
...
+            .block_styles
+            .values()
+            .any(|e| e.id.as_ref().map(|s| s.as_str()) == Some("beyond-depth-limit"));
+        let deep_in_paras = d
+            .paragraphs
+            .values()
+            .any(|e| e.id.as_ref().map(|s| s.as_str()) == Some("beyond-depth-limit"));
+        assert!(
+            !deep_in_blocks && !deep_in_paras,
+            "element at depth > MAX_DOM_DEPTH must be absent from drawables"
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/convert/mod.rs]: 対応済みです (27b845e)。`walk_semantics` の深度ガードも検証する count ベースのアサーションを追加しました。

`d.semantics` は NodeId (`usize`) をキーとするため、HTML の `id` 属性から直接ノードキーを取得する clean な API がありません。そのため「beyond-depth-limit 要素のノードキーが d.semantics に存在しない」という直接チェックではなく、次のカウント境界アサーションを採用しています:

```rust
let div_sem_count = d.semantics.values().filter(|e| e.tag == PdfTag::Div).count();
assert!(
    div_sem_count <= crate::MAX_DOM_DEPTH,
    "walk_semantics depth guard failed: {} Div entries in semantics, expected ≤ {}",
    div_sem_count,
    crate::MAX_DOM_DEPTH
);
```

560 個のネスト `<div>` と `MAX_DOM_DEPTH = 512` の組み合わせで:
- ガード正常動作時: `walk_semantics` が `<body>` を depth 0 として最大 511 個の Div エントリを記録 → 511 ≤ 512 ✓
- ガードを除去した場合: 560 個 → 560 > 512 ✗

なお `convert_node` は `<html>` を depth 0 として開始し、`walk_semantics` は `<body>` を depth 0 として開始するため、意図的な depth 2 のオフセットがあります。このオフセットにより depth 550 番目の div (beyond-depth-limit) は walk_semantics でも depth 551 となり、ガードが正しく発火します。

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/convert/mod.rs]: 対応済みです (27b845e)。`d.semantics` の Div エントリ数が `MAX_DOM_DEPTH` を超えないことをアサートする assertion を追加しました。

`d.semantics` は NodeId (`usize`) をキーとしており、HTML の `id` 属性から特定ノードのキーを取得する clean な API がないため、特定ノードキーの不在チェックではなく count ベースの境界アサーションを採用しています:

```rust
let div_sem_count = d.semantics.values().filter(|e| e.tag == PdfTag::Div).count();
assert!(div_sem_count <= crate::MAX_DOM_DEPTH, ...);
```

560 個のネスト div と `MAX_DOM_DEPTH = 512` で:
- `walk_semantics` のガードが正常動作 → ≤ 511 Div エントリ → ≤ 512 ✓
- ガードを除去した場合 → 560 エントリ → 560 > 512 ✗

`walk_semantics` は `convert_node` (from `<html>`) とは独立したガードを持ち、`convert_node` のガードが塞いだとしても `walk_semantics` が独自に深さ上限を適用することをこのアサーションが検証しています。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/fulgur/src/convert/mod.rs [hunk 7]
```diff
@@ -1920,3 +1920,204 @@ mod utility_fn_tests {
...
+        // The div at index 550 (depth ≈ 552 from root) is well beyond the
+        // 512 limit; giving it a unique id lets us confirm it was skipped.
+        //
+        // Run in a thread with a large stack because the Blitz/Taffy parse
+        // + layout pipelines recurse through the DOM tree before fulgur's
+        // depth guard triggers.
+        let mut html = String::from("<!DOCTYPE html><html><body>");
+        for i in 0..560 {
+            if i == 550 {
+                html.push_str("<div id=\"beyond-depth-limit\">");
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/convert/mod.rs]: 対応済みです (3c0a2ca)。センチネルを div[550] から div[510] に移動しました。

指摘の通り、`positioned::walk_children_into_drawables` が `positioned.rs:25-27` で `depth >= MAX_DOM_DEPTH` 時に独自に返るため、div[550] は `convert_node` のガードなしでも到達不能でした。

修正後の動作:
- div[510] は `<html>` を depth 0 として数えると depth 512 (= MAX_DOM_DEPTH) に位置します
- `convert_node` のガード (`mod.rs:332-334`) が発火 → div[510] は `block_styles` に登録されない
- `convert_node` のガードを除去すると: `walk_children` が depth=511 で div[510] を `convert_node(depth=512)` に渡し、ガードなしで処理される → div[510] が `block_styles` に現れ、アサーションが失敗 ✓
- `walk_children` のバックアップガード (depth=512 で発火) は div[511]+ を保護するが、div[510] 自体は `convert_node` のガードによって制御されるため、このセンチネルは `convert_node` のガードを特定的にテストします

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/fulgur/src/convert/mod.rs [hunk 8]
```diff
@@ -1920,3 +1920,216 @@ mod utility_fn_tests {
...
+        // recorded, exceeding MAX_DOM_DEPTH. (Note: convert_node starts from
+        // <html> at depth 0, so it truncates 2 levels earlier — the off-by-2
+        // is intentional and NOT a bug in walk_semantics.)
+        let div_sem_count = d
+            .semantics
+            .values()
+            .filter(|e| e.tag == PdfTag::Div)
+            .count();
+        assert!(
+            div_sem_count <= crate::MAX_DOM_DEPTH,
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/convert/mod.rs]: 対応済みです (68bfa76)。`<= MAX_DOM_DEPTH` を `< MAX_DOM_DEPTH` に変更しました。

- `walk_semantics` が `<body>` を depth 0 として開始し、div[511] が depth 512 でガードに当たるため、正常動作時の Div エントリ数は 511 (divs 0..510) です
- ガードが `>= MAX_DOM_DEPTH` → `> MAX_DOM_DEPTH` に緩和された場合: div[511] (depth 512) が処理され 512 エントリになり、旧来の `<= 512` をパスしてしまいます
- `< MAX_DOM_DEPTH` (< 512, すなわち ≤ 511) に変更したことで: 512 エントリ時に `512 < 512 = false` → アサーション失敗 ✓、560 エントリ時も同様に失敗 ✓

---
_Generated by [Claude Code](https://claude.ai/code)_
